crates/oxc_linter/src/rules/eslint/no_undef.rs

```diff
@@ -0,0 +1,191 @@
+use oxc_ast::AstKind;
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{Atom, Span};
+use oxc_syntax::operator::UnaryOperator;
+
+use crate::{context::LintContext, globals::BUILTINS, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-undef): Disallow the use of undeclared variables")]
+#[diagnostic(severity(warning), help("'{0}' is not defined."))]
+struct NoUndefDiagnostic(Atom, #[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoUndef {
+    #[allow(dead_code)]
+    type_of: bool,
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow the use of undeclared variables.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// It is most likely a potential ReferenceError caused by a misspelling of a variable or parameter name.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// var foo = someFunction();
+    /// var bar = a + 1;
+    /// ```
+    NoUndef,
+    correctness
+);
+
+impl Rule for NoUndef {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let type_of = value
+            .get(0)
+            .and_then(|config| config.get("typeof"))
+            .and_then(serde_json::Value::as_bool)
+            .unwrap_or_default();
+        Self { type_of }
+    }
+    fn run_once(&self, ctx: &LintContext) {
+        let symbol_table = ctx.symbols();
+
+        for reference_id_list in ctx.scopes().root_unresolved_references().values() {
+            for &reference_id in reference_id_list {
+                let reference = symbol_table.get_reference(reference_id);
+                if BUILTINS.contains_key(reference.name().as_str()) {
+                    return;
+                }
+                if !self.type_of {
+                    for node in ctx.semantic().nodes().iter() {
```

Comment:
Oh, this is not looking good.

We need to add `AstNodeId` to `Reference`. https://github.com/web-infra-dev/oxc/issues/685

Reply:
@makotot You may try adding `AstNodeId` to `Reference`, should be trivial (maybe).

Reply:
Thanks! I'll take a look.

Reply:
Fixed by https://github.com/web-infra-dev/oxc/pull/672/commits/5d16568bce3d1ab1736c26776dfdac0f8372da7b.